openhtf/plugs/frontline/bluetooth_sniffer.py

```diff
@@ -0,0 +1,277 @@
+      response = line.split(';')
+      _LOG.debug('Sniffer response: %s', line.strip())
+      if command in response and 'SUCCEEDED' in response:
+        return
+      raise SnifferError(full_command, line.strip())
+
+  def _export_tab(self, tab):
+    ending = '_' + tab.lower().rsplit(':', 1)[1].replace(' ', '_') + '.csv'
+    try:
+      self._sniffer_command('Export', 'File=' + self.export_prefix + ending,
+                            'Tab=' + tab)
+    except SnifferError as e:
+      _LOG.warning('Failed to export tab %s: %s', tab, e.response)
+    else:
+      self._files.append(self.export_prefix + ending)
+
+  @property
+  def files(self):
+    """Iterator over the names of log files created by the sniffer."""
+    for path in self._files:
+      yield os.path.basename(path)
+
+  @property
+  def export_prefix(self):
+    """Generate a filename prefix specific to a single sniffer run."""
+    return os.path.join(self._log_path, 'sniff_{}_{}'.format(
+        self.capture_start_time.strftime('%Y%m%d-%H%M%S'), self.capture_id))
+
+  def start_sniffing(self,
+                     bluetooth_mac,
+                     capture_id,
+                     binary_path=REMOTE_BIN_PATH):
+    """Configure the sniffer and start sniffing."""
+    self._connect()
+    assert self.capture_id is None
+    self.capture_id = capture_id
+    self.capture_start_time = datetime.datetime.now()
+
+    _LOG.info('Starting BLE sniffer')
+    self._sniffer_command('Start FTS', binary_path)
+    self._sniffer_command('Config Settings', 'IOParameters', 'FTSLE',
+                          'leDevice=0x' + bluetooth_mac.replace(':', ''))
+    time.sleep(10)  # Give enough time for FTS to detect sniffer and initialize.
+    self._sniffer_command('Start Sniffing')
+    self._sniffer_command('Start Capture')
+
+  def stop_sniffing(self):
+    """Stop the sniffer and save the collected data."""
+    if self.capture_id is None:
+      _LOG.warning('Attempted to stop sniffing when no capture is underway')
+      return
+
+    self._sniffer_command('Stop Capture')
+    self._sniffer_command('Save Capture', self.export_prefix + '.cfa')
+
+    self._files = []
+    self._files.append(self.export_prefix + '.cfa')
+    self._files.append(self.export_prefix + '.frm')
+    self._files.append(self.export_prefix + '.fsc')
+
+    self._export_tab('LE:LE LL')
+    self._export_tab('LE:LE DATA')
+
+    self._sniffer_command('Stop FTS')
+
+    _LOG.info('BLE data exported with prefix: %s', self.export_prefix)
+
+    self.reset()
+
+  def reset(self):
+    """Reset the Sniffer abstraction to a clean state."""
+    if self._socket:
+      self._socket.close()
+    self._socket = None
+    self.capture_id = None
+    self.capture_start_time = None
+
+  @contextlib.contextmanager
+  def sniff(self, bluetooth_mac, capture_id):
+    """Start and stop the sniffer through the context manager interface."""
+    self.start_sniffing(bluetooth_mac, capture_id)
+
+    try:
+      yield
```

Comment:
maybe yield self?  not sure it would actually be used at all, but it's a common paradigm for a contextmanager like this

Reply:
Done.